util/topsql/reporter/pubsub.go

```diff
@@ -0,0 +1,257 @@
+// Copyright 2021 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package reporter
+
+import (
+	"context"
+	"errors"
+	"time"
+
+	"github.com/pingcap/tidb/util"
+	"github.com/pingcap/tidb/util/logutil"
+	"github.com/pingcap/tipb/go-tipb"
+	"go.uber.org/atomic"
+	"go.uber.org/zap"
+)
+
+// TopSQLPubSubService implements tipb.TopSQLPubSubServer.
+//
+// If a client subscribes to TopSQL records, the TopSQLPubSubService is responsible
+// for registering an associated DataSink to the reporter. Then the reporter sends
+// data to the client periodically.
+type TopSQLPubSubService struct {
+	dataSinkRegHandle DataSinkRegHandle
+}
+
+// NewTopSQLPubSubService creates a new TopSQLPubSubService.
+func NewTopSQLPubSubService(
+	dataSinkRegHandle DataSinkRegHandle,
+) *TopSQLPubSubService {
+	return &TopSQLPubSubService{
+		dataSinkRegHandle: dataSinkRegHandle,
+	}
+}
+
+var _ tipb.TopSQLPubSubServer = &TopSQLPubSubService{}
+
+// Subscribe registers dataSinks to the reporter and redirects data received from reporter
+// to subscribers associated with those dataSinks.
+func (t *TopSQLPubSubService) Subscribe(
+	_ *tipb.TopSQLSubRequest,
+	stream tipb.TopSQLPubSub_SubscribeServer,
+) error {
+	ds := newPubSubDataSink(stream)
+	t.dataSinkRegHandle.Register(ds)
+	ds.run()
+	return nil
+}
+
+type pubSubDataSink struct {
+	stream     tipb.TopSQLPubSub_SubscribeServer
+	sendTaskCh chan sendTask
+	isDown     *atomic.Bool
+}
+
+func newPubSubDataSink(
+	stream tipb.TopSQLPubSub_SubscribeServer,
+) *pubSubDataSink {
+	return &pubSubDataSink{
+		stream:     stream,
+		sendTaskCh: make(chan sendTask),
+		isDown:     atomic.NewBool(false),
+	}
+}
+
+var _ DataSink = &pubSubDataSink{}
+
+func (s *pubSubDataSink) Send(data ReportData, deadline time.Time) {
+	if s.IsDown() {
+		return
+	}
+
+	select {
+	case s.sendTaskCh <- sendTask{data: data, deadline: deadline}:
+		// sent successfully
+	default:
```

Comment:
`sendTaskCh` is non-buffered, I'm worried that any concurrent calls to `Send()` will cause discarding. Should we use buffered channel?

Reply:
When will concurrent call happen?

Reply:
This is the description of `DataSink.Send`:
> If the sink is kept full and cannot schedule a send within the specified deadline, the data will be silently dropped.

Currently we only have one reportWorker, so no concurrent Send will occur. But the design of this API gives people the feeling that "I can call `Send` concurrently, and `DataSink` will maintain a reasonable threshold internally, and discarding occurs when the concurrency exceeds the threshold." However, the fact is that serial calls to `Send` will never "overload" , but whenever a parallel call is initiated, it will immediately "overload".

Reply:
Oh, good catch.

My intention was to buffer one `sendTask`. I am going to construct `sendTaskCh` with `make(chan sendTask, 1)`.